src/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -163,28 +163,61 @@ private void AddSolutionFolders(SlnFile slnFile, SlnProject slnProject)
                     "NestedProjects",
                     SlnSectionType.PreProcess);
 
+                var pathToGuidMap = GetSolutionFolderPaths(slnFile, nestedProjectsSection.Properties);
+
                 string parentDirGuid = null;
+                var solutionFolderHierarchy = string.Empty;
                 foreach (var dir in solutionFolders)
                 {
-                    var solutionFolder = new SlnProject
+                    solutionFolderHierarchy = Path.Combine(solutionFolderHierarchy, dir);
+                    if (pathToGuidMap.ContainsKey(solutionFolderHierarchy))
                     {
-                        Id = Guid.NewGuid().ToString("B").ToUpper(),
-                        TypeGuid = ProjectTypeGuids.SolutionFolderGuid,
-                        Name = dir,
-                        FilePath = dir
-                    };
-
-                    slnFile.Projects.Add(solutionFolder);
-
-                    if (parentDirGuid != null)
+                        parentDirGuid = pathToGuidMap[solutionFolderHierarchy];
+                    }
+                    else
                     {
-                        nestedProjectsSection.Properties[solutionFolder.Id] = parentDirGuid;
+                        var solutionFolder = new SlnProject
+                        {
+                            Id = Guid.NewGuid().ToString("B").ToUpper(),
+                            TypeGuid = ProjectTypeGuids.SolutionFolderGuid,
+                            Name = dir,
+                            FilePath = dir
+                        };
+
+                        slnFile.Projects.Add(solutionFolder);
+
+                        if (parentDirGuid != null)
+                        {
+                            nestedProjectsSection.Properties[solutionFolder.Id] = parentDirGuid;
+                        }
+                        parentDirGuid = solutionFolder.Id;
                     }
-                    parentDirGuid = solutionFolder.Id;
                 }
 
                 nestedProjectsSection.Properties[slnProject.Id] = parentDirGuid;
             }
         }
+
+        private IDictionary<string, string> GetSolutionFolderPaths(SlnFile slnFile, SlnPropertySet nestedProjects)
+        {
+            var solutionFolderPaths = new Dictionary<string, string>();
```

Comment:
Yes, and as long as these collections are populated consistently (no casing changes to the file names being added), it should work correctly.